e2e/framework.go

```diff
@@ -285,49 +351,27 @@ func (n *node) isStuck(num uint64) (uint64, bool) {
 	return 0, false
 }
 
-func (n *node) lastProposer() pbft.NodeID {
-	lastProposer := pbft.NodeID("")
-	if len(n.proposals) != 0 {
-		lastProposer = n.proposals[len(n.proposals)-1].Proposer
-	}
-	return lastProposer
-}
-
-func (n *node) getProposals() (uint64, []*pbft.SealedProposal) {
-	n.lock.Lock()
-	defer n.lock.Unlock()
-
-	res := []*pbft.SealedProposal{}
-	res = append(res, n.proposals...)
-
-	number := uint64(0)
-	if len(res) != 0 {
-		number = uint64(res[len(res)-1].Number)
-	}
-	return number, res
+func (n *node) Insert(pp *pbft.SealedProposal) error {
+	index := n.c.insertFinalProposal(pp)
+	n.setSyncIndex(index)
```

Comment:
I am not sure we need to set the index here, it would be better to do it at the end of `pbft.Run(ctx)`. At least in the `v3` repo, we change the current block after the protocol is over. Thus, we can also remove `insertFinalProposal` returning an index.

Reply:
Do you think we should update local index once the state machine goes to `done` state? Here sync index is updated once the proposal is finalised.

Reply:
Yes, update the index after `done` state. Then, if is enough to do `n.setSyncIndex(currentIndex+1)`.